packages/react-native/core/EventEmitter/NativeEventEmitter.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
+'use strict';
+
+import { type EventSubscription, type IEventEmitter } from '../vendor/emitter/EventEmitter';
+import Platform from '../Utilities/Platform';
+import RCTDeviceEventEmitter from './RCTDeviceEventEmitter';
+
+export interface NativeModule {
+  addListener(eventType: string): void;
+  removeListeners(count: number): void;
+}
+
+export type { EventSubscription };
+
+/**
+ * `NativeEventEmitter` is intended for use by Native Modules to emit events to
+ * JavaScript listeners. If a `NativeModule` is supplied to the constructor, it
+ * will be notified (via `addListener` and `removeListeners`) when the listener
+ * count changes to manage "native memory".
+ *
+ * Currently, all native events are fired via a global `RCTDeviceEventEmitter`.
+ * This means event names must be globally unique, and it means that call sites
+ * can theoretically listen to `RCTDeviceEventEmitter` (although discouraged).
+ */
+export class NativeEventEmitter implements IEventEmitter {
+  _nativeModule: NativeModule;
+
+  constructor(nativeModule: NativeModule) {
+    if (Platform.OS === 'ios') {
+      if (nativeModule === null) {
+        throw new Error('`new NativeEventEmitter()` requires a non-null argument.');
+      }
+    }
+
+    const hasAddListener =
+      // $FlowFixMe[method-unbinding] added when improving typing for this parameters
+      !!nativeModule && typeof nativeModule.addListener === 'function';
+    const hasRemoveListeners =
+      // $FlowFixMe[method-unbinding] added when improving typing for this parameters
+      !!nativeModule && typeof nativeModule.removeListeners === 'function';
+
+    if (nativeModule && hasAddListener && hasRemoveListeners) {
+      this._nativeModule = nativeModule;
+    } else if (nativeModule != null) {
+      if (!hasAddListener) {
+        console.warn('`new NativeEventEmitter()` was called with a non-null argument without the required `addListener` method.');
+      }
+      if (!hasRemoveListeners) {
+        console.warn('`new NativeEventEmitter()` was called with a non-null argument without the required `removeListeners` method.');
+      }
+    }
```

Comment:
Together with the above patch that assigns `nativeModule` to `this.nativeModule` implicitly, we could reduce this to:
```suggestion
    if (typeof nativeModule?.addListener !== 'function') {
      console.warn('`new NativeEventEmitter()` was called with a non-null argument without the required `addListener` method.');
    }
    if (typeof nativeModule?. removeListeners !== 'function') {
      console.warn('`new NativeEventEmitter()` was called with a non-null argument without the required `removeListeners ` method.');
    }
```
Once I understand how this is being used, I'd like to have a second pass at reviewing this part too, as generally we want to trust TypeScript as far as possible and avoid having to do type checks on construction args. The exception is when it's a bunch of dynamic stuff happening automatically that could bypass the checks (which this could well be).

Reply:
We probably need these two checks as we are not sure if those two methods exist on a native module or not. And I think these need to be added to our `NativeModuleHolder` class too.